sdks/java/extensions/ordered/src/main/java/org/apache/beam/sdk/extensions/ordered/OrderedEventProcessor.java

```diff
@@ -0,0 +1,657 @@
+      Iterable<TimestampedValue<EventTypeT>> events =
+          bufferedEventsState.readRange(startRange, endRange);
+
+      Iterator<TimestampedValue<EventTypeT>> bufferedEventsIterator = events.iterator();
+      while (bufferedEventsIterator.hasNext()) {
+        TimestampedValue<EventTypeT> timestampedEvent = bufferedEventsIterator.next();
+        Instant eventTimestamp = timestampedEvent.getTimestamp();
+        long eventSequence = eventTimestamp.getMillis();
+
+        EventTypeT bufferedEvent = timestampedEvent.getValue();
+        if (processingState.checkForDuplicateBatchedEvent(eventSequence)) {
+          outputReceiver
+              .get(unprocessedEventsTupleTag)
+              .output(
+                  KV.of(
+                      processingState.getKey(),
+                      KV.of(
+                          eventSequence,
+                          UnprocessedEvent.create(bufferedEvent, Reason.duplicate))));
+          continue;
+        }
+
+        if (eventSequence > processingState.getLastOutputSequence() + 1) {
+          processingState.foundSequenceGap(eventSequence);
+          // Records will be cleared up to this element
+          endClearRange = Instant.ofEpochMilli(eventSequence);
+          break;
+        }
+
+        // This check needs to be done after we checked for sequence gap and before we
+        // attempt to process the next element which can result in a new result.
+        if (exceededMaxResultCountForBundle(processingState, largeBatchEmissionTimer)) {
+          endClearRange = Instant.ofEpochMilli(eventSequence);
+          break;
+        }
+
+        state.mutate(bufferedEvent);
+        ResultTypeT result = state.produceResult();
+        if (result != null) {
+          outputReceiver.get(mainOutputTupleTag).output(KV.of(processingState.getKey(), result));
+          processingState.resultProduced();
+        }
+        processingState.processedBufferedEvent(eventSequence);
+        // Remove this record also
+        endClearRange = Instant.ofEpochMilli(eventSequence + 1);
+      }
+
+      bufferedEventsState.clearRange(startRange, endClearRange);
+    }
+
+    private boolean exceededMaxResultCountForBundle(
+        ProcessingState<EventKeyTypeT> processingState, Timer largeBatchEmissionTimer) {
+      boolean exceeded =
+          processingState.resultsProducedInBundle(numberOfResultsBeforeBundleStart)
+              >= maxNumberOfResultsToProduce;
+      if (exceeded) {
+        LOG.info(
+            "Setting the timer to output next batch of events for key '"
+                + processingState.getKey()
+                + "'");
+        // TODO: this work fine for global windows. Need to check what happens for other types of
+        // windows.
+        // See GroupIntoBatches for examples on how to hold the timestamp.
+        // TODO: test that on draining the pipeline all the results are still produced correctly.
```

Comment:
+1, would be good to verify both these conditions before we merge. I don't see a reason why this wouldn't work though

Reply:
Added tests to make sure that processing is done correctly in fixed windows. Added a link to a Beam task to test how the draining works.